src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -644,6 +622,50 @@ DeferErrorIfFromClauseRecurs(Query *queryTree)
 }
 
 
+/*
+ * DetermineQueryReturnsRecurringTuples returns tuple recurrence information
```

Comment:
Hmm, I think having `Query` in the function name is slightly confusing because we're not actually checking all parts of the query.

name suggestion: `FromClauseRecurringTupleType ()`
function comment suggestion: 
```
FromClauseRecurringTupleType recursively checks the query's range table list
 and returns a valid recurring tuple type if there are no distributed tables 
exist. Note that the function recursively checks all subqueries as well. 

If there are several candidate recurring tuples, we'll simply pick the first one.
```

Reply:
We are also interested in the recurrence type to provide a customized error message.  So I am not in favor for having a bool function. 

Agreed that it could have a better name